Dbeaver/dbeaver#20655 autocomplete multicolumn

serjiokov opened the pull request:
Provided approach of autocompletion multitable query with aliases.

serjiokov committed:
dbeaver/dbeaver#20655 Table autocomplete propose columns

Added autocompletioning for multitabel  definition in SQL Editor.

serjiokov committed:
dbeaver/dbeaver#20655 Added autocompletioning for multitabel aliases

serjiokov committed:
dbeaver/dbeaver#20655 Autocomplete support for multiple columns

serjiokov committed:
dbeaver/dbeaver#20655 Used NTLR for aliases identification

serjiokov committed:
dbeaver/dbeaver#20655 Extracted specific logic to method for code
transparency.

serjiokov commented:
![Screenshot 2023-09-27 183634](https://github.com/dbeaver/dbeaver/assets/32031248/365edda9-ede0-4f0a-a4b2-f2ef5d25528e)
![Screenshot 2023-09-27 183757](https://github.com/dbeaver/dbeaver/assets/32031248/3fa8793b-1cca-429b-ad1f-2f66ff707eae)

E1izabeth commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/completion/SQLCompletionAnalyzer.java:
```diff
@@ -1556,6 +1621,26 @@ public void visit(@Nullable CreateView createView) {
     }
+    
+    @NotNull
+    private String getAliceFromQueryNTLR(
```
?
```suggestion
    private String getAliasFromQueryANTLR(
```

E1izabeth commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/completion/SQLCompletionAnalyzer.java:
```diff
@@ -937,6 +948,39 @@ private void makeDataSourceProposals(@NotNull Map<String, Object> parameters) th
     }
 
+    @Nullable
+    private List<DBSObject> getTableListFromAlias(DBSObjectContainer sc, @Nullable String token)
```
There is no sense of having two almost identical methods. 
getTableFromAlias does the same thing, let's remove it and refactor usages to use this method returning list.

Or maybe it's better to refactor out a common part to a separate method.

serjiokov replied:
I suggest to keep as is for now, touching this method invokes waterfall changes in code and tests. BMHO better to plan refatoring task to cleanUp this class.

serjiokov committed:
dbeaver/dbeaver#20655 Method name fixed

E1izabeth commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/completion/SQLCompletionAnalyzer.java:
```diff
@@ -1556,6 +1621,26 @@ public void visit(@Nullable CreateView createView) {
+        @NotNull String tableName,
+        @NotNull String query) throws IOException {
+        String alias = "";
+        STMSource querySource = STMSource.fromReader(new StringReader(query));
+        LSMAnalyzer analyzer = LSMDialectRegistry.getInstance().getAnalyzerForDialect(request.getContext().getDataSource().getSQLDialect());
+        STMTreeRuleNode tree = analyzer.parseSqlQueryTree(querySource, new STMSkippingErrorListener());
```
It's not a good idea to call parser more than once for one query. Reuse the analyzer please (this.tableRefsAnalyzer)

serjiokov replied:
Rewoked to use a reference implemenatation.

serjiokov committed:
dbeaver/dbeaver#20655 Added test for completion support

E1izabeth commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/completion/SQLCompletionAnalyzer.java:
```diff
@@ -1532,7 +1535,8 @@ public void visit(@Nullable CreateView createView) {
                         // It is table name completion after FROM. Auto-generate table alias
                         SQLDialect sqlDialect = SQLUtils.getDialectFromObject(object);
-                        alias = SQLUtils.generateEntityAlias((DBSEntity) object, s -> {
+                        DBSEntity object22 = (DBSEntity) object;
+                        alias = SQLUtils.generateEntityAlias(object22, s -> {
```
Eliminate redundant variable or give it meaningful name

E1izabeth commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/completion/SQLCompletionAnalyzer.java:
```diff
@@ -1495,16 +1497,18 @@ private SQLCompletionProposalBase makeProposalsFromObject(
+            SQLConstants.KEYWORD_JOIN.equals(prevWord)) {
             if (object instanceof DBSEntity) {
-                aliasMode = SQLTableAliasInsertMode.fromPreferences(((DBSEntity) object).getDataSource().getContainer().getPreferenceStore());
+                DBSEntity object2 = (DBSEntity) object;
+                aliasMode = SQLTableAliasInsertMode
+                    .fromPreferences(object2.getDataSource().getContainer().getPreferenceStore());
```
Eliminate redundant variable or give it meaningful name

E1izabeth commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/completion/SQLCompletionAnalyzer.java:
```diff
@@ -1622,24 +1640,28 @@ public void visit(@Nullable CreateView createView) {
+     * The method returns table to alias map
+     */
+    @Nullable
+    private Map<String, String> getTableAlicesFromQueryByANTLRParser(
```
We have LSMTableReferencesAnalyzer interface with 2 implementations. One implementation contains method for getting table references from syntax tree built by antlr parser, another one is previously used analyzer. And implementation is regulated by the setting Enable experimental feature for autocompletion. Let's keep this idea and move this method to one of the realization of this interface (TableReferencesAnalyzer). Method for OldTableReferencesAnalyzer could be empty if there was nothing previously for this case. Also it will allow to turn this feature off if something goes wrong with parser.

serjiokov replied:
Provided reference implementation, extrcated to specific packages and classes.

serjiokov committed:
dbeaver/dbeaver#20655 Extracted reference implementation SQL analyzers.

Provided tests covers completion from multiple tables with aliases.

serjiokov committed:
dbeaver/dbeaver#20655 Fixes for static and style checks

serjiokov committed:
dbeaver/dbeaver#20655 Fixes for style checks

serjiokov commented:
Test cases passed successfully.
![image](https://github.com/dbeaver/dbeaver/assets/32031248/1a321e70-75bc-4bc3-8a0d-9bfdba1bf527)

E1izabeth commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/analyzer/internal/impl/TableReferencesAnalyzerImpl.java:
```diff
@@ -0,0 +1,131 @@
+        this.request = request;
+    }
+
+    private void prepareTableReferences() {
+        final SQLScriptElement activeQuery = request.getActiveQuery();
+        if (activeQuery == null) {
+            tableReferences = Collections.emptyMap();
+            return;
+        }
```
Make this method to accept `String query` and move `if (activeQuery == null)` to the caller. Then you can call it from `getTableAlicesFromQuery`. Otherwise we parse the same query multiple times.

serjiokov replied:
No objections, good advice.

E1izabeth commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/analyzer/internal/impl/TableReferencesAnalyzerImpl.java:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    @Override
+    public Map<String, String> getTableAlicesFromQuery(String query) {
+        Map<String, String> table2alias = new TreeMap<>();
+        try {
+            STMSource querySource = STMSource.fromReader(new StringReader(query));
+            final LSMAnalyzer analyzer = LSMDialectRegistry.getInstance()
+                .getAnalyzerForDialect(request.getContext().getDataSource().getSQLDialect());
+            STMTreeRuleNode tree = analyzer.parseSqlQueryTree(querySource, new STMSkippingErrorListener());
+            table2alias = getTableAndAliasFromSources(tree);
+        } catch (Exception e) {
+            log.error(e.getMessage(), e);
+        }
+        return table2alias;
```
```suggestion
    public Map<String, String> getTableAliasesFromQuery(String query) {
        try {
            prepareTableReferences(query);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
        return tableReferences;
```

serjiokov replied:
Done.

serjiokov committed:
dbeaver/dbeaver#20655 Supported completion after WHERE...AND...
keywords.

serjiokov commented:
Autocompletion support added for multiple tables with/without aliases.
Use cases on screens bellow:


![Screenshot 2023-09-27 183757](https://github.com/dbeaver/dbeaver/assets/32031248/affde79c-3599-43fc-af6e-da504db65e5a)
![Screenshot 2023-09-29 172602](https://github.com/dbeaver/dbeaver/assets/32031248/22f13650-74f6-46c8-8568-20c2ae754dfb)
![Screenshot 2023-09-29 172819](https://github.com/dbeaver/dbeaver/assets/32031248/8d301e28-e6f0-4aac-b361-7c28a4e36400)
![Screenshot 2023-09-29 172912](https://github.com/dbeaver/dbeaver/assets/32031248/7c6b3f04-bac0-42c5-b0b6-e512540b59e1)
![Screenshot 2023-09-29 172948](https://github.com/dbeaver/dbeaver/assets/32031248/460ff66c-f9ac-4830-817d-90c99be29d13)
![Screenshot 2023-09-29 173230](https://github.com/dbeaver/dbeaver/assets/32031248/acac8ac4-9990-4bfe-a128-a22770286048)

serjiokov committed:
dbeaver/dbeaver#20655 Fixes after review

serge-rider commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/analyzer/TableReferencesAnalyzer.java:
```diff
@@ -0,0 +1,51 @@
+     * The method returns map of table by alias base on query expression
+     */
+    @NotNull
+    Map<String, String> getTableAlicesFromQuery(@NotNull String query);
```
Alices -> aliases

serjiokov replied:
Method name fixed.

serge-rider commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/analyzer/internal/impl/InlineState.java:
```diff
@@ -0,0 +1,26 @@
+ */
+package org.jkiss.dbeaver.model.sql.analyzer.internal.impl;
+
+public enum InlineState {
```
Make it private inner class. It never used outside.

serge-rider commented on plugins/org.jkiss.dbeaver.model.sql/src/org/jkiss/dbeaver/model/sql/analyzer/internal/impl/InlineState.java:
```diff
@@ -0,0 +1,26 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.model.sql.analyzer.internal.impl;
```
No need to create packages internal. All completion stuff is internal. So just move all classes to org.jkiss.dbeaver.model.sql.analyzer

serjiokov replied:
The package internal name is design in eclsipe comunity to avoid inheritance or usage calsses in such package.

serjiokov replied:
The implentation parts are always internal, interfaces are not.

serjiokov committed:
dbeaver/dbeaver#20655 Fixes after review

Moved implementation to ...sql.analyzer package
Moved InlineState enum to private class

E1izabeth commented:
For QA: **Please, check how it works with `Enable experimental feature` off and on**

E1izabeth committed:
Merge branch 'devel' into dbeaver/dbeaver#20655-autocomplete-multicolumn

MashaKorax commented:
verified
